src/preset.ts

```diff
@@ -445,17 +600,53 @@ export async function expandConfigurePreset(folder: string,
 
   expansionOpts.envOverride = expandedPreset.environment as EnvironmentVariables;
 
+  const presetsFile = getPresetsFile(folder);
+  if (presetsFile && presetsFile.version >= 3) {
+    // For presets v3+ binaryDir and generator are optional, but cmake-tools needs a value. Default to something reasonable.
+    if (!preset.binaryDir) {
+      const defaultValue =  '${sourceDir}/out/build/${presetName}';
+
+      log.warning(localize('binaryDir.undefined', 'Configure preset {0}: No binaryDir specified, using default value "{1}"', preset.name, defaultValue));
+      preset.binaryDir = defaultValue;
+    }
+    if (!preset.generator) {
+      const defaultValue = 'Ninja';
```

Comment:
The CMake docs say:
*In version 3 or above, this field may be omitted to fall back to regular generator discovery procedure.*

Are we supposed to emulate that or is there some strategy that will allow our generate command to match what might happen on the CLI?

Reply:
I'm not aware of any documented strategy for determining the generator. The closest I found was from [CMAKE_GENERATOR](https://cmake.org/cmake/help/latest/envvar/CMAKE_GENERATOR.html#envvar:CMAKE_GENERATOR)
> If the provided value doesn't name a generator known by CMake, the internal default is used.

which is pretty unhelpful. I went with `Ninja` just because that's what we do in VS. I can see what happens if I leave generator undefined, but I think there are a few places that assume generator is present. We could try matching CMake behavior experimentally too but, at least to me, it seems like more trouble than it's worth.

Reply:
VS has the benefit of knowing that Ninja is always available.  Maybe we could check to see if the `cmake.generator` or `cmake.preferredGenerators` setting is set and use that as the first choice.

Reply:
Good point, I changed it to try falling back to those settings first before falling back to Ninja.